### PR TITLE
Unbundle-able libraries

### PR DESCRIPTION
**Pull request by QuLogic**
This moves bundled libraries into a single directory, and adds shims so that the system copies _may_ be used instead. `six` was a bit of a pain, so I pulled the work that was in astropy/astropy#2623. This should make it easier for packagers, while still leaving things self-contained for you.

### CODE REVIEW

File: vispy/ext/decorator.py [hunk 1]
```diff
@@ -1,253 +1,14 @@
...
-    ContextManager = type(
-        'ContextManager', (GeneratorContextManager,),
-        dict(__call__=__call__, __init__=__init__))
+from __future__ import absolute_import
 
-contextmanager = decorator(ContextManager)
+try:
+    from vispy.ext._bundled.decorator import *
+except ImportError:
+    from decorator import *
```
Comment (Eric89GXL) [inline vispy/ext/decorator.py]: but the above should never fail, right?

File: vispy/ext/decorator.py [hunk 2]
```diff
@@ -1,253 +1,14 @@
...
-    def __init__(self, f, *a, **k):
-        return GeneratorContextManager.__init__(self, f(*a, **k))
-    ContextManager = type(
-        'ContextManager', (GeneratorContextManager,),
-        dict(__call__=__call__, __init__=__init__))
+from __future__ import absolute_import
 
-contextmanager = decorator(ContextManager)
+try:
+    from vispy.ext._bundled.decorator import *
```
Comment (Eric89GXL) [inline vispy/ext/decorator.py]: ...and we prefer to use relative imports, why not do from .decorator import * here?

**Comment by Eric89GXL** [comment]
What problem does this fix? I'm still not clear on that.

**Comment by QuLogic** [comment]
Packages don't like bundled libraries. Before, you'd have to change all references to `vispy.ext.*`, but now, you can just delete the `_bundled` directory, and add package dependencies as usual.

**Comment by Eric89GXL** [comment]
By packagers you mean like Anaconda? I'm missing who this will help

**Comment by QuLogic** [comment]
[Fedora](https://fedoraproject.org/wiki/Packaging:No_Bundled_Libraries), [openSUSE](https://en.opensuse.org/openSUSE:Packaging_guidelines#Bundling_of_multiple_projects), Debian (can't find a link, but there are [cases](https://wiki.debian.org/Teams/RustPackaging/Unbundling)), etc.

File: vispy/ext/decorator.py (hunk 2 above)
Comment (QuLogic) [inline vispy/ext/decorator.py]: If you wish, but I think you mean `from ._bundled.decorator import *`?

**Comment by campagnola** [comment]
I ran into the same issues with pyqtgraph. Even had a hard time packaging the documentation because of a single .js file that sphinx dragged in.

**Comment by Eric89GXL** [comment]
Ahh makes sense. To properly test support, we should modify one of our Travis runs (probably 3.4 full) so that the external dependencies are installed as well, and then also `vispy/ext/_bundled` gets deleted before any tests are run. @QuLogic can you add that to this PR? It will test that the unbundling actually works as expected.

**Comment by Eric89GXL** [comment]
Also @QuLogic do you have particular bundling targets in mind that you (or others) plan on working on?

**Comment by QuLogic** [comment]
Well, technically speaking, all bundled libraries should be removed, but practically speaking, I think that's all I'm up for at the moment. Of the items in `ext`:
 * `cubehelix` hasn't made any releases,
 * `egl` appears home-grown,
 * `fontconfig` could maybe be unbundled, but it would mean a hard dependency on pyglet, which I'd rather avoid if possible (haven't looked into it yet),
 * `gdi32plus` is not a big deal here,
 * `glfw` does not seem to have an upstream,
 * `ipy_inputhook` maybe, but IPython would be a heavy dep, and looking at other tickets, it looks like it might get modified/simplified even more.

Maybe could look at this one:
 * `freetype` says "adapted from freetype-py", so I'd have to look into the changes first, but there may already be packages for it.

I assume everything bundled is in `ext`.

**Comment by Eric89GXL** [comment]
We have tried to keep bundled things in there, yes. That sounds reasonable to me, thanks for taking a stab at this @QuLogic. I think there is a `pyglw` or `glfwpy`, but I'm not sure what types of releases they've had. Probably safer to leave that one.

Do you have time to modify Travis appropriately as I commented above?

**Comment by Eric89GXL** [comment]
I think it would mean a conditional in the `.travis.yml` checking if the run is Python 3.4 full, and if so, delete the `ext/_bundle` directory, and use `conda` and if not available then `pip` to install the packages.

**Comment by QuLogic** [comment]
OK, I'll look into it in a little bit, then.

**Comment by Eric89GXL** [comment]
Thanks -- it should ensure that the code all works as expected, and ensure that we keep pace with any changes in those packages as we go.

**Comment by QuLogic** [comment]
Freetype seems to be causing trouble somehow. Maybe I shouldn't have added that one...

**Comment by QuLogic** [comment]
Looks like [that was commented out](https://github.com/vispy/vispy/blob/master/vispy/ext/freetype.py#L192-201). I think a better fix is upstream rougier/freetype-py#21, but it's not in a released version yet.

**Comment by rougier** [comment]
I'll try to make a new release. Can you open an issue on freetype-py (just to remind me of making a release).

**Comment by QuLogic** [comment]
OK, I left out freetype-py for now, so that this will work. I'll hold on to that change until freetype-py gets a new release.

**Comment by Eric89GXL** [comment]
Thanks a lot @QuLogic for getting Travis on board

**Comment by QuLogic** [comment]
That... seems a little late there...

**Comment by Eric89GXL** [comment]
Coveralls does that sometimes, guess he gets lazy

**Pull request merged**